### ISSUE TITLE
dataclass instance variable is unassignable to same type when typed as Self | None

### ISSUE THREAD

**Issue by blhsing** [microsoft/pyright#10042]
The following snippet:
```python
from typing import Self
from dataclasses import dataclass

@dataclass
class Node:
    next_node: Self | None = None

    def foo(self):
        self.next_node = Node()
```
produces the complaint:
```
Cannot assign to attribute "next_node" for class "Node*"
  Type "Node" is not assignable to type "Self@Node | None"
    Type "Node" is not assignable to type "Self@Node"
    "Node" is not assignable to "None"  (reportAttributeAccessIssue)
```
However, the equivalent code without dataclass does not get any complaint:
```python
from typing import Self

class Node:
    def __init__(self, next_node: Self | None = None):
        self.next_node = next_node

    def foo(self):
        self.next_node = Node()
```
Demo [here](https://pyright-play.net/?pythonVersion=3.14&locale=en-us&code=GYJw9gtgBALgngBwJYDsDmUkQWEMoDKApgDbACwAUKJFACYCGMDAxiQwM4dEeba75GzNpw5UqAASGt2XKiK5QAcmDpEAXFSjaoKIgA8YAfRSqNhUsCgAfZWD1QAvHb3jKO%2BkSvAwYABTcZACUmu4eOoHAAHR6hiZmTnZqfkFAA)

Pyright version: 1.1.396
Python version: 3.14

Note that the first snippet does not get a complaint from mypy.

**Comment by erictraut** [comment]
Pyright is working correctly here and in conformance with the Python typing spec.

The `Self` special form is an implicit type variable that represents the type of the `self` parameter.

Consider the case where you subclass `Node`. The type of `self` would be a type other than `Node`. You would therefore be violating a typing rule by assigning an instance of `None` to `next_node`. So pyright is correct to emit an error here.

You can fix this type violation in your code by changing the `foo` method to the following:

```python
    def foo(self):
        self.next_node = type(self)()
```

**Comment by blhsing** [comment]
@erictraut Thanks for the explanation.

However, I believe your interpretation of what `Self` represents is quite different from Python's official [documentation](https://docs.python.org/3/library/typing.html#typing.Self), which states that `Self` is a "Special type to represent the current enclosed class", rather than the type of the `self` parameter as you suggested.

So `self.next_node = Node()` should be valid even in a subclass because `next_node` is typed equivalent to `TypeVar("Self", bound="Node")`.

If your interpretation were true then I suppose we should file a bug against the behavior of the second snippet.

Thanks in advance for your reconsideration.

**Comment by erictraut** [comment]
The general Python documentation is incorrect here — or is at least misleading.

The authoritative documentation you should refer to for typing functionality is the Python typing spec, which can be found [here](https://typing.python.org/en/latest/spec/index.html). The section that talks about `Self` can be found [here](https://typing.python.org/en/latest/spec/generics.html#self).

Pyright is a standards-compliant type checker, and it passes all of the type checker conformance tests, including those that are related to `Self`. Search for "generics_self" on [this page](https://htmlpreview.github.io/?https://github.com/python/typing/blob/main/conformance/results/results.html) for a list of relevant compliance tests.

If you want to refer to a specific class within your code, you should use the name of that class. For example, you could modify your above sample as follows, and this will type check without error.

```python
from typing import Self
from dataclasses import dataclass

@dataclass
class Node:
    next_node: "Node | None" = None

    def foo(self):
        self.next_node = Node()
```

The `Self` special form was added to the type system to specifically handle subclassing. It is not simply a shorthand for the containing class.

In your second example, you have not provided any type declaration for the instance variable `next_node`, so pyright allows you to assign any value to it from within the class implementation. So pyright is working as designed here as well.

BTW, if you have general questions about the Python type system, the [Python typing forum](https://discuss.python.org/c/typing/32) is a good place to post those questions. It is monitored by typing experts.

**Comment by blhsing** [comment]
@erictraut Ah you're absolutely correct. The first sentence of `typing.Self` section of the general documentation may be misleading but the rest of it describes its behavior in a subclass in details just like you suggested. Both the typing spec and PEP-673 have pointed it out too. And you're absolutely right about the missing annotation for `self.next_node` in my second snippet.

Many thanks for taking the time to bring me up to speed. I'm now off to filing a bug against mypy for not complaining about my first snippet. ;-)
